compiler/src/dotty/tools/dotc/CompilationUnit.scala

```diff
@@ -85,6 +85,18 @@ class CompilationUnit protected (val source: SourceFile) {
   def assignmentSpans(using Context): Map[Int, List[Span]] =
     if myAssignmentSpans == null then myAssignmentSpans = Nullables.assignmentSpans
     myAssignmentSpans
+
+  /** Force pickling at end of Pickler phase for determinism
+   *
+   *  On Windows we observe non-deterministic failure for the
+   *  idempotency test tests/pos/i6507b.scala. The cause of the
+   *  non-determinism is that the positions of the to be pickled
+   *  trees might change in a later phase `Inlining` after pickling.
```

Comment:
@nicolasstucki Could Inlining avoid changing tree positions? Otherwise we could copy the tree first, this should still be more efficient than forcing pickling early.

Reply:
Inlining is not supposed to change positions, `Inlined` nodes are supposed to have the same position as the original tree. Could be a bug setting the positions or maybe it is constant folding that removes some positions.

Reply:
Even if we fix the bug in `inlining`, it's so easy to re-introduce the bug in the future.

Or, we add an assertion in `def span_=` to check that spans are only set once. It seems to be a reasonable invariant.